LifetimeDependenceScopeFixup: fix handling of returned dependence

atrick opened the pull request:
LifetimeDependenceScopeFixup: fix handling of returned dependence

Only rewrite the mark_dependence to depend on the function argument when the
dependent value is actually returned.

Also, find all uses even if an escaping use is seen.

Add full support for local variables, including flow sensitive analysis.

atrick committed:
LifetimeDependence: clarify log headers

atrick committed:
LifetimeDependenceScopeFixup: fix handling of returned dependence

Only rewrite the mark_dependence to depend on the function argument when the
dependent value is actually returned.

Also, find all uses even if an escaping use is seen.

atrick committed:
SwiftCompilerSources: make BasicBlock hashable

atrick committed:
OwnershipLiveness: handle ProjectBox like an interior pointer.

atrick committed:
LifetimeDependenceUtils: handle borrowed ProjectBox

atrick committed:
Add NonEscapingClosureDefUseWalker

To find all effective applications of partial_apply [on-stack].

atrick committed:
LifetimeDependenceScopeFixup: handle non-dominated dependent uses.

atrick committed:
LifetimeDependenceDiagnostics: find the correct variable introducer.

atrick committed:
Add LocalVariableUtils

This includes a reaching-def analysis that can be used by diagnostic passes. Diagnostics should be flow-sensitive.

atrick committed:
LifetimeDependenceDefUseWalker: use LocalVariableReachableUses.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LifetimeDependenceScopeFixup.swift:
```diff
@@ -40,93 +40,110 @@ let lifetimeDependenceScopeFixupPass = FunctionPass(
-    // for nested access scopes
-    guard case let .scope(enclosingBeginAccess) = beginAccess.address.enclosingAccessScope else {
-      break
+  // Create new end_access at the end of extended uses
```
Does it re-create end accesses even if the scope already fits?

atrick replied:
Right, I think that can be avoided, just didn't get around to it yet:
https://github.com/apple/swift/pull/71703/commits/c17643cf0285cd52efd3c805aa6f6df76328a5b1

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LifetimeDependenceScopeFixup.swift:
```diff
@@ -40,93 +40,110 @@ let lifetimeDependenceScopeFixupPass = FunctionPass(
+    if let arg = extendAccessScopes(dependence: lifetimeDep, context) {
+      markDep.baseOperand.set(to: arg, context)
     }
-    extendDependenceBase(dependenceInstruction: markDep,
```
Is this not needed anymore?

atrick replied:
`extendDependenceBase` was a hack to handle the special case in which the dependent value is returned from a function and the access scope is in the caller. This case is handled directly now so we don't need the hack.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LifetimeDependenceScopeFixup.swift:
```diff
@@ -40,93 +40,110 @@ let lifetimeDependenceScopeFixupPass = FunctionPass(
+  // Create new end_access at the end of extended uses
+  for end in range.ends {
+    let endBuilder = Builder(after: end, context)
+    _ = endBuilder.createEndAccess(beginAccess: beginAccess)
```
`Builder.createEndAccess` should have a `@discardableResult` attribute, because `end_access` has no result value.

eeckstein reviewed (commented):
Looks fine to me, though I didn't think through all the details. @meg-gupta can probably do a better review than me.

atrick committed:
SwiftCompilerSources: add Value.nextInstruction

atrick committed:
Review feedback; Builder.createEndAccess()

atrick committed:
Review feedback; only rewrite access scopes when necessary

atrick committed:
Update tests for precise lifetime diagnostics.

atrick committed:
Redesign gatherBorrowIntroducers to return BeginBorrowValues.

This avoids a lot of confusion because the callers expect this type. Fixing it just required some redundancy and
bridging in the EnclosigValues implementation.

atrick committed:
Add SingleInlineArray read/_modify accessors

For mutable access.

atrick committed:
SwiftCompilerSources: Add Instruction.endInstruction

To make the ScopedInstruction abstraction useful and formalize to use-points of scoped operations.

This is specifically for scoped operations that use their operands up to the end of their scope.

This is not for the many instructions that introduce scoped lifetimes. Allocations are not considered scoped
operations. Every owned value has a scoped lifetime just like an allocation, but the scope-ending instructions are not
considered use points.

atrick committed:
Factor InstructionRange.beginningInstruction for use elsewhere.

atrick committed:
Add AddressOwnershipLiveRange

A live range representing the ownership of addressible memory.

This live range represents the minimal guaranteed lifetime of the object being addressed. Uses of derived addresses
may be extended up to the ends of this scope without violating ownership.

.liveOut objects (@in_guaranteed, @out and globals) have no instruction range.

.local objects (alloc_stack, yield, @in, @inout) report the single live range of the full assignment that reaches
this address.

.owned values (boxes and references) simply report OSSA liveness.

.borrow values report each borrow scope's range. The effective live range is their intersection. A valid use must
lie within

atrick committed:
Add computeBorrowLiveRange

Compute the live range for the borrow scopes of a guaranteed value. This returns a separate instruction range for
each of the value's borrow introducers. Unioning those ranges would be incorrect. We typically want their
intersection.

atrick commented:
@swift-ci test

atrick committed:
Increase BlockCapacity from 8 to 16 for intrusive BasicBlockSet.

InstructionRange uses 3 sets. Some algorithms need two active ranges at the top-level. Additionally, utilities often have few levels of nesting that each require a block set.

atrick commented:
@swift-ci smoke test
